File: scripts/merge-toml-config.js
const fs = require("fs");
const path = require("path");

const repoConfigPath = process.argv[2];
const userConfigPath = process.argv[3];

if (!repoConfigPath || !userConfigPath) {
    console.error("Usage: node merge-toml-config.js <source-config> <target-config>");
    process.exit(1);
}

// Read shared disabled list
const sharedDisabledPath = path.join(__dirname, "..", "shared-disabled-mcp.json");
let sharedDisabled = [];
try {
    const shared = JSON.parse(fs.readFileSync(sharedDisabledPath, "utf8"));
    sharedDisabled = shared.disabledMcpServers || [];
} catch (e) {
    // No shared config, that's fine
}

function parseTOML(content) {
    const lines = content.split(/\r?\n/);
    const root = {};
    let currentTable = null;

    for (let line of lines) {
        line = line.trim();
        if (!line || line.startsWith("#")) {
            continue;
        }

        const tableMatch = line.match(/^\[([^\]]+)\]$/);
        if (tableMatch) {
            currentTable = tableMatch[1].trim();
            if (!root[currentTable]) {
                root[currentTable] = {};
            }
            continue;
        }

        const kvMatch = line.match(/^([^=]+)=\s*(.+)$/);
        if (kvMatch) {
            const key = kvMatch[1].trim();
            const val = kvMatch[2].trim();
            if (currentTable) {
                root[currentTable][key] = val;
            } else {
                root[key] = val;
            }
        }
    }
    return root;
}

function stringifyTOML(obj) {
    let result = "";

    for (const key of Object.keys(obj)) {
        if (typeof obj[key] !== "object") {
            result += `${key} = ${obj[key]}\n`;
        }
    }
    result += "\n";

    for (const key of Object.keys(obj)) {
        if (typeof obj[key] === "object") {
            result += `[${key}]\n`;
            for (const subKey of Object.keys(obj[key])) {
                result += `${subKey} = ${obj[key][subKey]}\n`;
            }
            result += "\n";
        }
    }
    return result;
}

try {
    const repoContent = fs.readFileSync(repoConfigPath, "utf8");
    const userContent = fs.readFileSync(userConfigPath, "utf8");

    const repoObj = parseTOML(repoContent);
    const userObj = parseTOML(userContent);

    const mergedObj = { ...userObj };

    // Remove disabled mcp_servers from both repo and user configs
    for (const name of sharedDisabled) {
        const serverKey = `mcp_servers.${name}`;
        delete repoObj[serverKey];
        delete mergedObj[serverKey];
    }

    // Remove URL-based mcp_servers from repo (Codex only supports stdio)
    for (const key of Object.keys(repoObj)) {
        if (key.startsWith("mcp_servers.") && typeof repoObj[key] === "object" && repoObj[key].url) {
            delete repoObj[key];
        }
    }

    for (const key of Object.keys(repoObj)) {
        if (typeof repoObj[key] === "object") {
            if (!mergedObj[key]) {
                mergedObj[key] = { ...repoObj[key] };
            } else {
                if (key === "mcp_servers" || key.startsWith("mcp_servers.")) {
                    // For mcp_servers, we always prefer repo's definition to fix broken/outdated command/args
                    mergedObj[key] = { ...mergedObj[key], ...repoObj[key] };
                } else {
                    // Merge keys inside table, preferring user's existing keys
                    mergedObj[key] = { ...repoObj[key], ...mergedObj[key] };
                }
            }
        } else {
            if (mergedObj[key] === undefined) {
                mergedObj[key] = repoObj[key];
            }
        }
    }

    const mergedContent = stringifyTOML(mergedObj);
    fs.writeFileSync(userConfigPath, mergedContent, "utf8");
    console.log("Successfully merged Codex configuration into " + userConfigPath);
} catch (err) {
    console.error("Failed to merge Codex config:", err.message);
    process.exit(1);
}

File: scripts/merge_toml_config.py
#!/usr/bin/env python3
import sys
import json
import re
from pathlib import Path

def parse_toml(content: str) -> dict:
    lines = content.splitlines()
    root = {}
    current_table = None

    for line in lines:
        line = line.strip()
        if not line or line.startswith("#"):
            continue

        table_match = re.match(r"^\[([^\]]+)\]$", line)
        if table_match:
            current_table = table_match.group(1).strip()
            if current_table not in root:
                root[current_table] = {}
            continue

        kv_match = re.match(r"^([^=]+)=\s*(.+)$", line)
        if kv_match:
            key = kv_match.group(1).strip()
            val = kv_match.group(2).strip()
            if current_table:
                root[current_table][key] = val
            else:
                root[key] = val
    return root


def stringify_toml(obj: dict) -> str:
    result = []
    
    # First write top-level primitive keys
    for key, val in obj.items():
        if not isinstance(val, dict):
            result.append(f"{key} = {val}")
    result.append("")

    # Then write tables
    for key, val in obj.items():
        if isinstance(val, dict):
            result.append(f"[{key}]")
            for sub_key, sub_val in val.items():
                result.append(f"{sub_key} = {sub_val}")
            result.append("")
            
    return "\n".join(result) + "\n"


def main():
    if len(sys.argv) < 3:
        print("Usage: python merge_toml_config.py <source-config> <target-config>", file=sys.stderr)
        sys.exit(1)

    repo_config_path = Path(sys.argv[1])
    user_config_path = Path(sys.argv[2])

    # Read shared disabled list
    shared_disabled_path = Path(__file__).resolve().parent.parent / "shared-disabled-mcp.json"
    shared_disabled = []
    if shared_disabled_path.exists():
        try:
            shared = json.loads(shared_disabled_path.read_text(encoding="utf-8"))
            shared_disabled = shared.get("disabledMcpServers", [])
        except Exception:
            pass

    try:
        repo_content = repo_config_path.read_text(encoding="utf-8")
        user_content = user_config_path.read_text(encoding="utf-8")

        repo_obj = parse_toml(repo_content)
        user_obj = parse_toml(user_content)

        merged_obj = dict(user_obj)

        # Remove disabled mcp_servers from both repo and user configs
        for name in shared_disabled:
            server_key = f"mcp_servers.{name}"
            repo_obj.pop(server_key, None)
            merged_obj.pop(server_key, None)

        # Remove URL-based mcp_servers from repo (Codex only supports stdio)
        for key in list(repo_obj.keys()):
            if key.startswith("mcp_servers.") and isinstance(repo_obj[key], dict) and "url" in repo_obj[key]:
                repo_obj.pop(key, None)

        # Clean up agents in user config that are no longer registered in repo config
        for key in list(merged_obj.keys()):
            if key.startswith("agents.") and key not in repo_obj:
                merged_obj.pop(key, None)

        for key, val in repo_obj.items():
            if isinstance(val, dict):
                if key not in merged_obj:
                    merged_obj[key] = dict(val)
                else:
                    if key == "mcp_servers" or key.startswith("mcp_servers."):
                        # For mcp_servers, we always prefer repo's definition to fix broken/outdated command/args
                        merged_obj[key] = {**merged_obj[key], **val}
                    else:
                        # Merge keys inside table, preferring user's existing keys
                        merged_obj[key] = {**val, **merged_obj[key]}
            else:
                if key not in merged_obj:
                    merged_obj[key] = val

        merged_content = stringify_toml(merged_obj)
        user_config_path.write_text(merged_content, encoding="utf-8")
        print(f"Successfully merged Codex configuration into {user_config_path}")
    except Exception as err:
        print(f"Failed to merge Codex config: {err}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
